[TEP-0044]: Propose pipeline in pod + rollout plan

lbernick opened the pull request:
This commit proposes using Pipeline in a Pod to address TEP-0044,
and adds implementation milestones for features that should be supported in a pod.
Additional design details will be added in a subsequent commit.

lbernick commented on teps/0044-data-locality-and-pod-overhead-in-pipelines.md:
```diff
@@ -188,6 +188,7 @@ some functionality would require changes to this code, and some functionality ma
 * Functionality that could be supported with current pod logic (e.g. by
   [translating a Pipeline directly to a TaskRun](#pipeline-executed-as-taskrun)):
   * Sequential tasks (specified using [`runAfter`](https://github.com/tektoncd/pipeline/blob/main/docs/pipelines.md#using-the-runafter-parameter))
+  * [Parallel tasks](https://github.com/tektoncd/pipeline/blob/main/docs/pipelines.md#configuring-the-task-execution-order)
```
I've moved parallel tasks to "changes that won't require updating entrypoint" because the entrypoint already supports multiple waitfiles; it would just be a matter of updating the controller logic to determine what Tasks a given Task depends on based on a DAG.

bobcatfish replied:
great catch!! i was definitely overcomplicating it in my head 🤣

jerop commented:
/assign

lbernick commented:
/kind tep

bobcatfish commented:
/assign

bobcatfish commented on teps/0044-data-locality-and-pod-overhead-in-pipelines.md:
```diff
@@ -301,11 +301,75 @@ controller to make this decision.
+such as data from previous Tasks contained in Workspaces.
+- **Parallel Tasks**. Another primary use case for this TEP is to support parallel Tasks that share data, which today typically must be run in separate pods on the same node.
+Running parallel Tasks on the same pod wouldn't allow computation to be balanced between nodes, but it would allow users to avoid paying performance and UX penalties
+associated with PVCs or other workspace backing options. In addition, `finally` Tasks, which are also a required feature, run in parallel.
```
good call re. finally tasks + the parallel requirement 👍

bobcatfish commented on teps/0044-data-locality-and-pod-overhead-in-pipelines.md:
```diff
@@ -301,11 +301,75 @@ controller to make this decision.
+
+- **Pipeline- or Task-level resource requirements**: Users may be more likely to care about the resource usage of the pod rather than individual containers,
+since Kubernetes uses the pod’s resource requirements to make decisions on scheduling and eviction.
+- **Pipeline-level sidecars**: This feature is requested in [issue 4235](https://github.com/tektoncd/pipeline/issues/4235). 
```
maybe i missed it above but probably makes sense to include task level sidecars in the list above

i'm thinking optional but probably not that hard to add?

okay with it being required too - ive been thinking that it would be useful to be able to define a task with just a sidecar that can be easily composed with other tasks (e.g. to provide a docker in docker sidecar) - however that's a bit outside the scope of how this TEP is currently defined

lbernick replied:
good call-- I've added it for now to optional features.

bobcatfish reviewed (commented):
Looks great to me!

/approve

afrittoli commented on teps/0044-data-locality-and-pod-overhead-in-pipelines.md:
```diff
@@ -301,11 +301,75 @@ controller to make this decision.
+
+- **`when` expressions**
+- **retries**
+- **timeouts**
```
Does this refer to task timeouts or the whole pipeline timeout or both?
I think having at least the overall pipeline timeout is rather important, I would not want to run pipelines in a production environment if there's no control against them running indefinitely.
If timeouts are especially challenging, we could at least document available workarounds

lbernick replied:
just FYI I did end up adding a milestone for these two features!

afrittoli commented on teps/0044-data-locality-and-pod-overhead-in-pipelines.md:
```diff
@@ -301,11 +301,75 @@ controller to make this decision.
+- **`when` expressions**
+- **retries**
+- **timeouts**
+- **Matrixed Tasks (fanned out from params)**
```
I think this is very important as well. In terms of timeline I agree this should come after the required features, but I think it would be sad not to have it - especially since most of the heavy lifting will be done by having support for parallel tasks and workspaces.
Of course the fallback option here is to write the tasks one by one and run the resulting pipeline in a pod :)

lbernick replied:
makes sense! discussion in previous comment.

afrittoli commented on teps/0044-data-locality-and-pod-overhead-in-pipelines.md:
```diff
@@ -301,11 +301,75 @@ controller to make this decision.
+#### Unsupported Features
+
+The following features are likely infeasible, regardless of how useful they may be.
+- **Matrixed Tasks (fanned out from results)**: See [Dynamically created TaskRuns in Pipelines](#dynamically-created-taskruns-in-pipelines).
```
Sadly yes, but having static matrix would already be a great plus I think!

afrittoli commented on teps/0044-data-locality-and-pod-overhead-in-pipelines.md:
```diff
@@ -581,8 +624,8 @@ Pros:
 Cons:
-* Will need to update our entrypoint logic to allow for containers running in parallel
-  * We could (at least initially) only support sequential groups
+* No reason to introduce a new grouping syntax when we plan to have [Pipelines in Pipelines](./0056-pipelines-in-pipelines.md).
```
We don't have it yet in the core api - but hopefully soon enough :)

jerop replied:
yes, pipelines in pipelines is coming very soon 😁

afrittoli reviewed (commented):
Looks good, thank you!

I still would prefer we moved away from the "pipeline in a pod" naming of the feature - at least in the API and user docs - since the Pod is the mean and not the objective and the `Pod` is very k8s specific, but not something for this PR.

Looking forward to this, thank you @lbernick for driving it forward

/approve

vdemeester commented:
Same remarks as @afrittoli but looks good to me otherwise 😉

lbernick commented:
> I still would prefer we moved away from the "pipeline in a pod" naming of the feature - at least in the API and user docs - since the Pod is the mean and not the objective and the `Pod` is very k8s specific, but not something for this PR.

Updated to note that the naming of this feature is still under discussion!

jerop commented on teps/0044-data-locality-and-pod-overhead-in-pipelines.md:
```diff
@@ -154,8 +154,8 @@ We can explore providing this functionality in a later iteration of this proposa
-1. Any configuration for the execution of a Pipeline must be modifiable at runtime.
-  - We may explore adding authoring time configuration in the future after gathering feedback on runtime configuration.
+1. The chosen solution should allow us to provide both authoring time and runtime configuration to colocate multiple Tasks.
+  - We will start with only configuration for one of (authoring time, runtime) and optionally provide the other based on user feedback.
```
sounds good - though would prefer that we make this decision (authoring time vs runtime) before marking this TEP implementable

jerop commented on teps/0044-data-locality-and-pod-overhead-in-pipelines.md:
```diff
@@ -301,11 +301,78 @@ controller to make this decision.
+Users will choose which parts of a pipeline to run in a pod by grouping them into a sub-Pipeline, using the [Pipelines in Pipelines](./0056-pipelines-in-pipelines.md) feature.
+Any Pipelines grouped under a Pipeline executed in a pod will also execute in that pod.
+
+We will maintain the 1-1 relationship between Tasks and TaskRuns by creating one TaskRun per Pipeline Task and using it as the source of truth for the status of the
+execution of that Task.
```
this changes though with [matrixed pipelinetasks](https://github.com/tektoncd/community/blob/main/teps/0090-matrix.md), maybe what you wanted to say this?

```suggestion
We will create TaskRun(s) per PipelineTask and use them as the source of truth for the status of the execution of that PipelineTask.
```

jerop replied:
this changes though with [matrixed pipelinetasks](https://github.com/tektoncd/community/blob/main/teps/0090-matrix.md), maybe what you wanted to say this?

```suggestion
We will create TaskRun(s) per PipelineTask and use them as the source of truth for the status of the execution of that PipelineTask.
```

jerop commented on teps/0044-data-locality-and-pod-overhead-in-pipelines.md:
```diff
@@ -301,11 +301,78 @@ controller to make this decision.
+The following features are useful for PipelineRuns executed in pods for the same reasons that they are useful in PipelineRuns
+executed in multiple pods. Not all of these features will be equally easy to implement in a pod.
+
+- **`when` expressions**
+- **retries**
+- **timeouts**
+- **task-level sidecars**
+- **Matrixed Tasks (fanned out from params)**
+
```
could you please add some rationale on why these features are optional? also, can we reframe this as features that will be supported in future work instead of optional?

jerop commented on teps/0044-data-locality-and-pod-overhead-in-pipelines.md:
```diff
@@ -154,8 +154,8 @@ We can explore providing this functionality in a later iteration of this proposa
   - This requirement is being included because we could choose a solution that doesn't address the above use case.
-1. Any configuration for the execution of a Pipeline must be modifiable at runtime.
-  - We may explore adding authoring time configuration in the future after gathering feedback on runtime configuration.
+1. The chosen solution should allow us to provide both authoring time and runtime configuration to colocate multiple Tasks.
+  - We will start with only configuration for one of (authoring time, runtime) and optionally provide the other based on user feedback.
```
sounds good - though would prefer that we make this decision (authoring time vs runtime) before marking this TEP `implementable`

lbernick replied:
definitely agree! the goal of this PR is mainly to get consensus on pipeline in a pod as a preferred solution, and I think authoring time vs runtime will be discussed in more detail when we get to the API changes.

jerop commented on teps/0044-data-locality-and-pod-overhead-in-pipelines.md:
```diff
@@ -301,11 +301,78 @@ controller to make this decision.
+The following features are useful for PipelineRuns executed in pods for the same reasons that they are useful in PipelineRuns
+executed in multiple pods. Not all of these features will be equally easy to implement in a pod.
+
+- **`when` expressions**
+- **retries**
+- **timeouts**
+- **task-level sidecars**
+- **Matrixed Tasks (fanned out from params)**
```
could you please add some rationale on why these features are optional? also, can we reframe this as features that will be supported in future work instead of optional?

lbernick replied:
renamed optional -> future work! I didn't provide rationale for individual features, but I'm seeing this as a cost benefit tradeoff since this set of features isn't directly related to the data locality problem. I suspect that for matrix (for example), the benefit is high, and the implementation effort is lower, while retries will be much more challenging to implement so we may want to determine how much benefit we're providing. I didn't provide rationale for individual features as the execution mode isn't really relevant to how useful they are (just how difficult they are to implement), but I added a bit more detail to the description of the section.

jerop replied:
thanks for updating it to future work and expounding on rationale, much clearer :)

jerop reviewed (commented):
looks good, thanks @lbernick! 

left some comments/questions about authoring vs runtime, mapping btwn pipelinetasks and taskruns, and optional features

lbernick commented:
thanks @jerop, should be ready for another look!

pritidesai commented:
API WG - @jerop to lgtm after reviewing

jerop reviewed (approved):
looks great, thanks @lbernick 🎉

jerop commented:
as agreed upon at API WG on 03/14, lgtm'ing offline after approvals from all assignees

/lgtm
